extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchProcessor.java

```diff
@@ -0,0 +1,204 @@
+                first = true;
+                devService = null;
+                cfg = null;
+            };
+            closeBuildItem.addCloseTask(closeTask, true);
+        }
+        cfg = configuration;
+
+        if (devService.isOwner()) {
+            log.infof(
+                    "Dev Services for Elasticsearch started. Other Quarkus applications in dev mode will find the "
+                            + "server automatically. For Quarkus applications in production mode, you can connect to"
+                            + " this by starting your application with -D%s=%s",
+                    hostsConfigProperty, getElasticsearchHosts(hostsConfigProperty));
+        }
+        return devService.toBuildItem();
+    }
+
+    public static String getElasticsearchHosts(String hostsConfigProperty) {
+        return devService.getConfig().get(hostsConfigProperty);
+    }
+
+    private void shutdownElasticsearch() {
+        if (devService != null) {
+            try {
+                devService.close();
+            } catch (Throwable e) {
+                log.error("Failed to stop the Elasticsearch server", e);
+            } finally {
+                devService = null;
+            }
+        }
+    }
+
+    private DevServicesResultBuildItem.RunningDevService startElasticsearch(ElasticsearchDevServicesBuildTimeConfig config,
+            DevservicesElasticsearchBuildItem devservicesElasticsearchBuildItem,
+            LaunchModeBuildItem launchMode, boolean useSharedNetwork, Optional<Duration> timeout) throws BuildException {
+        if (!config.enabled.orElse(true)) {
+            // explicitly disabled
+            log.debug("Not starting dev services for Elasticsearch, as it has been disabled in the config.");
+            return null;
+        }
+
+        String hostsConfigProperty = devservicesElasticsearchBuildItem.getHostsConfigProperty();
+        // Check if elasticsearch hosts property is set
+        if (ConfigUtils.isPropertyPresent(hostsConfigProperty)) {
+            log.debugf("Not starting dev services for Elasticsearch, the %s property is configured.", hostsConfigProperty);
+            return null;
+        }
+
+        if (!isDockerWorking.getAsBoolean()) {
+            log.warnf(
+                    "Docker isn't working, please configure the Elasticsearch hosts property (%s).", hostsConfigProperty);
+            return null;
+        }
+
+        // Hibernate search Elasticsearch have a version configuration property, we need to check that it is coherent
+        // with the image we are about to launch
+        if (devservicesElasticsearchBuildItem.getVersionConfigProperty() != null) {
+            String version = devService.getConfig().get(devservicesElasticsearchBuildItem.getVersionConfigProperty());
+            String containerTag = config.imageName.substring(config.imageName.indexOf(':') + 1);
+            if (!containerTag.startsWith(devservicesElasticsearchBuildItem.getVersionConfigProperty())) {
```

Comment:
```suggestion
            if (!containerTag.startsWith(version)) {
```

But even then, the "version" could be prefixed with `opensearch:` or `elastic:`, in which case the recommendation in the error message would be weird. I suspect we'll need a more complex solution to make this work, like providing a parsing function in the build item (not just the property key, but also a function that extracts the version from the property key). Opensearch will be another problem, but one I'd gladly move to a separate issue.

Reply:
I'll switch to the provider of the build item to set both version and distribution and to limit devservices to the elastic distribution for now.

Reply:
I created #23906 to address OpenSearch dev-services.